File: 13-roman-to-integer/roman-to-integer.py
class Solution(object):

    symbols = {"I": 1, "V": 5, "X": 10, "L": 50, "C": 100, "D": 500, "M": 1000}

    def romanToInt(self, s):
        """
        :type s: str
        :rtype: int
        """
        total = 0
        chars = list(s) # characters
        size = len(chars)

        for i in range(size):
            letter = chars[i] # current roman symbol
            value = Solution.symbols[letter] # value of current roman symbol
            
            sflag = 1   # sign flag
            if i < size - 1: # bound check
                next_letter = chars[i+1]
                if Solution.symbols[next_letter] > value: # is the next value greater? 
                    sflag = -1 # flip the sign
            total += sflag * value
        return total
            
        
        

File: 13-roman-to-integer/roman-to-integer.java
class Solution {

    private static final Map<Character, Integer> SYMBOL_MAP = new HashMap<Character, Integer> () {{
        put('I', 1);
        put('V', 5);
        put('X', 10);
        put('L', 50);
        put('C', 100);
        put('D', 500);
        put('M', 1000);
    }};

    public int romanToInt(String s) {
        int total = 0;

        for (int i=0, j=1; i < s.length(); i++, j++) {
            int value = SYMBOL_MAP.get(s.charAt(i)); // current roman symbol value
            int sflag = 1; // sign flag

            if (j < s.length()) {
                int next_value = SYMBOL_MAP.get(s.charAt(j)); // next roman symbol value
                if (next_value > value) {
                    sflag = -1;
                }
            }

            total += sflag * value;
        }
        
        return total;
    }
}